src/accountsdb/cache.zig

```diff
@@ -0,0 +1,331 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Slot = sig.core.time.Slot;
+const Pubkey = sig.core.pubkey.Pubkey;
+
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+
+/// Stores read-only in-memory copies of commonly used accounts
+pub const AccountsCache = struct {
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        // this account has since been mutated, and will not progress to newer slots
+        // TODO: when we start mutating accounts, make sure to set this field
+        is_dirty: std.atomic.Value(bool) = .{ .raw = false },
+        ref_count: std.atomic.Value(usize),
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) !CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = std.atomic.Value(usize).init(1),
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            const current_count = self.ref_count.load(.acquire);
+            if (current_count == 1) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            } else {
+                _ = self.ref_count.fetchSub(1, .acq_rel);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            _ = self.ref_count.fetchAdd(1, .acq_rel);
+            return self;
+        }
+    };
+
+    pub const LRU = LruCacheCustom(
+        .locking,
+        Pubkey,
+        *CachedAccount,
+        std.mem.Allocator,
+        CachedAccount.releaseOrDestroyDoublePtr,
+    );
+    pub const SlotLRUs = std.AutoHashMapUnmanaged(Slot, LRU);
+
+    const Self = @This();
+
+    slot_lrus: SlotLRUs,
+    allocator: std.mem.Allocator,
+    max_items: usize,
+    max_slots: usize,
+    highest_slot: ?Slot,
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        max_items: usize,
+        max_slots: usize,
+    ) !AccountsCache {
+        return .{
+            .slot_lrus = .{},
+            .allocator = allocator,
+            .max_items = max_items,
+            .max_slots = max_slots,
+            .highest_slot = null,
+        };
+    }
+
+    pub fn get(self: *const Self, slot: Slot, pubkey: Pubkey) ?CachedAccount {
```

Comment:
What's the point of returning a `CachedAccount` here, if not as a a pointer, and without `copyRef`? This seems like a potential source of race conditions. Please correct me if I'm off the mark here.

Reply:
https://github.com/Syndica/sig/pull/297/commits/ae36aa27390e13793a692a6c0f2c1356439b4e27#diff-7f62328c8d0fd51e206b91ff2b72aef34de0fae176ba46106a0c768d29f306f8L86-R104